### PR TITLE
Add Kotlin/JVM support

### PR DESCRIPTION
**Pull request by DanielBelow**
This adds support for Kotlin/JVM.

Info on getting a standalone Kotlin compiler can be found on [kotlinlang.org](https://kotlinlang.org/docs/command-line.html), e.g. from [Kotlin's github releases](https://github.com/JetBrains/kotlin/releases/tag/v1.4.32). 
Note not to use the `kotlin-native-*` archives, but e.g. `kotlin-compiler-1.4.32.zip`.

### CODE REVIEW

File: etc/config/kotlin.defaults.properties [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Default settings for Kotlin/JVM
+compilers=&kotlin
+compilerType=kotlin
+versionFlag=-version
+objdumper=javap
+instructionSet=java
+
+group.kotlin.compilers=kotlincdefault
+compiler.kotlincdefault.exe=/usr/bin/kotlinc
```
Comment (supergrecko) [inline etc/config/kotlin.defaults.properties:9]: We are not interested in allowing kotlinc-js to be invoked. The kotlinc-jvm driver should be used for this.

```suggestion
compiler.kotlincdefault.exe=/usr/bin/kotlinc-jvm
```

File: lib/compilers/_all.js [hunk 2]
```diff
@@ -41,6 +41,7 @@ export { GCCRSCompiler } from './gccrs';
 export { GolangCompiler } from './golang';
 export { HaskellCompiler } from './haskell';
 export { ISPCCompiler } from './ispc';
+export { KotlinCompiler } from './kotlin';
 export { JavaCompiler } from './java';
```
Comment (supergrecko) [inline lib/compilers/_all.js:45]: These exports should be alphabetically ordered

```suggestion
export { ISPCCompiler } from './ispc';
export { JavaCompiler } from './java';
export { KotlinCompiler } from './kotlin';
```

**Review by supergrecko — CHANGES_REQUESTED** [review]
Looks good. However, there are a few things we should consider changing and one thing to add.

The kotlinc/kotlinc-jvm drivers take different arguments than the Java Compiler and therefore we should filter out different arguments.

We are aiming to look at the generated bytecode from the Kotlin source file which means we should be preventing KotlinScript from running. I believe all of the following arguments should be filtered out in filterUserOptions

- -script
- -progressive
- -d
- -kotlin-home
- -jdk-home-path

**Comment by DanielBelow** [comment]
Thanks for being so quick to review:)

I've added the arguments you suggested, but I noticed there's also `-Xjavac-arguments=<options>`, to forward arguments to `javac`, so I might be able to pass options to `javac` this way, that would be filtered out by `java.js`'s `filterUserOptions`(?)

A similar flag exists for NVCC (`-Xcompiler`), but I don't see any special handling of that in `nvcc.js`.

Should we also exclude `-Xjavac-arguments`?

**Comment by supergrecko** [comment]
Yes, I think we should exclude that as well. Can simply just filter anything that starts with -Xjavac

File: lib/compilers/kotlin.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2019, Compiler Explorer Authors
```
Comment (RubenRBS) [inline lib/compilers/kotlin.js:1]: Outdated copyright year
```suggestion
// Copyright (c) 2021, Compiler Explorer Authors
```

**Review by RubenRBS — APPROVED** [review]

**Comment by partouf** [comment]
Can you describe the installation process? Or could you write it in a similar way in which we install java for example https://github.com/compiler-explorer/infra/blob/main/bin/yaml/java.yaml

**Comment by DanielBelow** [comment]
I linked to the documentation on how to install in my first comment: https://kotlinlang.org/docs/command-line.html
For a manual install:

> Unzip the standalone compiler into a directory and optionally add the bin directory to the system path. The bin directory contains the scripts needed to compile and run Kotlin on Windows, OS X and Linux.

Standalone compiler can be downloaded from the github releases: https://github.com/JetBrains/kotlin/releases/tag/v1.4.32, so the direct link to a standalone v1.4.32 version is https://github.com/JetBrains/kotlin/releases/download/v1.4.32/kotlin-compiler-1.4.32.zip

I _assume_ it could work like the java example, of course checking for `bin/kotlinc-jvm -version` instead:

```
compilers:
  kotlin:
    compression: zip
    dir: kotlin-{name}
    check_exe: bin/kotlinc-jvm -version
    targets:
      - name: 1.4.32
        url: https://github.com/JetBrains/kotlin/releases/download/v1.4.32/kotlin-compiler-1.4.32.zip
```

**Comment by supergrecko** [comment]
One thing we should ensure when setting up infra for Kotlin/JVM is to make it easy for us to enable Kotlin/Native further down the road. Maybe we should prefix dir with `kotlin-jvm-{name}` assuming the dir key is where the binaries will be stored.

**Comment by supergrecko** [comment]
Applying the following patch for the `kotlin.amazon.properties` seems to make it work with the compilers added in https://github.com/compiler-explorer/infra/pull/510

https://gist.github.com/supergrecko/7ff0b687d2f58c4a188a3ac1f62b8c20

**Review by supergrecko — APPROVED** [review]
LGTM. Tested this locally. There's definitely some filtering work to do with the javap output (such as LocalVariableTable showing up), but I aim to improve that further down the road as the current implementation is pretty hacky. 
image

**Comment by partouf** [comment]
Something I encountered during installation and checking afterwards:
The scripts `kotlinc-jvm` and `kotlinc` require a `java` in the path or `JAVA_HOME` to be set. These are not available by default on any of our instances.

Not entirely sure what's the best option, but likely something with setting JAVA_HOME to one of the java's we have installed. For example `/opt/compiler-explorer/jdk-16.0.1`. We would either need to set this inside of CE, or on our instances.

I would think inside CE would be best, as that's more flexible and wouldn't require us to make new images of our instances.
A global JAVAHOME (or something) property would probably be fine, but per kotlin compiler would probably be best.

**Comment by supergrecko** [comment]
> I would think inside CE would be best, as that's more flexible and wouldn't require us to make new images of our instances.
> A global JAVAHOME (or something) property would probably be fine, but per kotlin compiler would probably be best.

A per Kotlin compiler solution seems like the ideal solution here. Setting a global JAVA_HOME could also cause problems further down the road if we introduce other Java-based tools.

**Comment by DanielBelow** [comment]
@partouf Just so I understand correctly: this would look similar to what golang is doing when looking up `goroot`, `goarch` and `goos`? 
Something like ``this.compilerProps(`compiler.${this.compiler.id}.java_home`);``, with java_home being defined in the `*.properties` files?

**Comment by supergrecko** [comment]
I believe calling compilerProps with 'java_home' when there's a java_home entry present should suffice.

**Comment by DanielBelow** [comment]
The way I understood the discussion was that a per-compiler solution was best.
Can we have per-compiler `JAVA_HOME`s, if we query for `java_home` (without compiler-id) here?

I had something like this in mind:
```
getDefaultExecOptions() {
    const execOptions = super.getDefaultExecOptions();
    const javaHome = this.compilerProps(`compiler.${this.compiler.id}.java_home`);
    if (javaHome) {
        execOptions.env.JAVA_HOME = javaHome;
    }

    return execOptions;
}
```
so that kotlin-jvm v1.4.3 could have a different java home from kotlin-jvm v1.5, but maybe I misunderstand how `compilerProps` queries the info?

**Comment by partouf** [comment]
> The way I understood the discussion was that a per-compiler solution was best.
> Can we have per-compiler `JAVA_HOME`s, if we query for `java_home` (without compiler-id) here?
> 
> I had something like this in mind:
> 
> ```
> getDefaultExecOptions() {
>     const execOptions = super.getDefaultExecOptions();
>     const javaHome = this.compilerProps(`compiler.${this.compiler.id}.java_home`);
>     if (javaHome) {
>         execOptions.env.JAVA_HOME = javaHome;
>     }
> 
>     return execOptions;
> }
> ```
> 
> so that kotlin-jvm v1.4.3 could have a different java home from kotlin-jvm v1.5, but maybe I misunderstand how `compilerProps` queries the info?

This is in principle correct, but I would move the `const javaHome = this.compilerProps(`compiler.${this.compiler.id}.java_home`);` thing to here https://github.com/compiler-explorer/compiler-explorer/blob/main/lib/compiler-finder.js#L249

That way you don't have to query the props every compilation request, and can ask for `this.compiler.javaHome` in `getDefaultExecOptions()`

**Comment by partouf** [comment]
Can you also add the java_home properties to the amazon.properties? You can see where they are installed in the java amazon prop file

**Comment by partouf** [comment]
I think this is at a stage where its good to go on testing on the real thing. I believe I have installed Kotlin a while ago already.

I will work on it tomorrow probably.

**Comment by partouf** [comment]
Testing on https://godbolt.org/staging  - seems we forgot about a language thing, will try to remember.. and fix

**Comment by partouf** [comment]
@DanielBelow @supergrecko 

https://godbolt.org/staging/

Kotlin compilers working:

- 1.4.0
- 1.4.10
- 1.5.0

Not working:

- 1.4.20
- 1.4.21
- 1.4.30
- 1.4.31
- 1.4.32


Ideas?

**Pull request merged**